File: internal/models/service.go
// agency-gateway/internal/models/service.go
package models

import (
	"fmt"
	"strings"
	"unicode"
)

// ServiceCredentialConfig defines how a service authenticates.
type ServiceCredentialConfig struct {
	EnvVar       string  `yaml:"env_var" validate:"required"`
	Header       string  `yaml:"header" validate:"required"`
	Format       *string `yaml:"format"`
	ScopedPrefix string  `yaml:"scoped_prefix" validate:"required"`
}

// Validate checks scoped_prefix starts with "agency-scoped-".
func (s *ServiceCredentialConfig) Validate() error {
	if !strings.HasPrefix(s.ScopedPrefix, "agency-scoped-") {
		return fmt.Errorf("scoped_prefix must start with 'agency-scoped-'")
	}
	return nil
}

// ServiceToolParameter defines a parameter for a service tool.
type ServiceToolParameter struct {
	Name        string  `yaml:"name" validate:"required"`
	Type        string  `yaml:"type" default:"string"`
	Description string  `yaml:"description" validate:"required"`
	Required    bool    `yaml:"required" default:"true"`
	Default     *string `yaml:"default"`
}

type ConsentRequirement struct {
	OperationKind    string `yaml:"operation_kind"`
	TokenInputField  string `yaml:"token_input_field"`
	TargetInputField string `yaml:"target_input_field"`
	MinWitnesses     int    `yaml:"min_witnesses"`
}

// ServiceTool defines an MCP-exposed tool for a service.
type ServiceTool struct {
	Name                 string                 `yaml:"name" validate:"required"`
	Description          string                 `yaml:"description" validate:"required"`
	Scope                string                 `yaml:"scope"`
	Parameters           []ServiceToolParameter `yaml:"parameters"`
	Method               string                 `yaml:"method" default:"GET"`
	Path                 string                 `yaml:"path" validate:"required"`
	QueryParams          map[string]string      `yaml:"query_params"`
	BodyTemplate         map[string]interface{} `yaml:"body_template"`
	ResponsePath         *string                `yaml:"response_path"`
	RequiresConsentToken *ConsentRequirement    `yaml:"requires_consent_token"`
}

// ServiceDefinition is a service that can be granted to agents.
type ServiceDefinition struct {
	Service      string                  `yaml:"service" validate:"required"`
	DisplayName  string                  `yaml:"display_name" validate:"required"`
	APIBase      string                  `yaml:"api_base" validate:"required"`
	Description  string                  `yaml:"description"`
	Author       string                  `yaml:"author,omitempty"`
	License      string                  `yaml:"license,omitempty"`
	Credential   ServiceCredentialConfig `yaml:"credential"`
	UsageExample *string                 `yaml:"usage_example"`
	Tools        []ServiceTool           `yaml:"tools"`
}

// Validate checks service name is alphanumeric with hyphens/underscores,
// and validates the credential config.
func (s *ServiceDefinition) Validate() error {
	for _, r := range s.Service {
		if !unicode.IsLetter(r) && !unicode.IsDigit(r) && r != '-' && r != '_' {
			return fmt.Errorf("service name must be alphanumeric with hyphens or underscores")
		}
	}
	if err := s.Credential.Validate(); err != nil {
		return err
	}
	for _, tool := range s.Tools {
		if tool.RequiresConsentToken == nil {
			continue
		}
		if tool.RequiresConsentToken.OperationKind == "" {
			return fmt.Errorf("tool %q requires_consent_token.operation_kind is required", tool.Name)
		}
		if tool.RequiresConsentToken.TokenInputField == "" {
			return fmt.Errorf("tool %q requires_consent_token.token_input_field is required", tool.Name)
		}
		if tool.RequiresConsentToken.TargetInputField == "" {
			return fmt.Errorf("tool %q requires_consent_token.target_input_field is required", tool.Name)
		}
		params := make(map[string]bool, len(tool.Parameters))
		for _, param := range tool.Parameters {
			params[param.Name] = true
		}
		if !params[tool.RequiresConsentToken.TokenInputField] {
			return fmt.Errorf("tool %q requires_consent_token references unknown token_input_field %q", tool.Name, tool.RequiresConsentToken.TokenInputField)
		}
		if !params[tool.RequiresConsentToken.TargetInputField] {
			return fmt.Errorf("tool %q requires_consent_token references unknown target_input_field %q", tool.Name, tool.RequiresConsentToken.TargetInputField)
		}
	}
	return nil
}

// ServiceGrant records that a service was granted to an agent.
type ServiceGrant struct {
	Service       string   `yaml:"service" validate:"required"`
	GrantedAt     string   `yaml:"granted_at" validate:"required"`
	GrantedBy     string   `yaml:"granted_by" validate:"required"`
	AllowedScopes []string `yaml:"allowed_scopes,omitempty"`
}

// Validate checks granted_at is not empty.
func (g *ServiceGrant) Validate() error {
	if g.GrantedAt == "" {
		return fmt.Errorf("granted_at must not be empty")
	}
	return nil
}

// AgentServiceGrants holds all service grants for an agent.
type AgentServiceGrants struct {
	Agent  string         `yaml:"agent" validate:"required"`
	Grants []ServiceGrant `yaml:"grants"`
}

File: images/models/service.py
"""Service credential models.

Defines the schema for service definitions (what services exist and how they
authenticate) and service grants (which agents have access to which services).
"""

from datetime import datetime, timezone

from pydantic import BaseModel, ConfigDict, field_validator
from typing import Optional


class ServiceCredentialConfig(BaseModel):
    """How a service authenticates API requests."""

    model_config = ConfigDict(extra="forbid")

    env_var: str
    header: str
    format: Optional[str] = None
    scoped_prefix: str

    @field_validator("scoped_prefix")
    @classmethod
    def prefix_must_start_with_agency(cls, v: str) -> str:
        if not v.startswith("agency-scoped-"):
            raise ValueError("scoped_prefix must start with 'agency-scoped-'")
        return v


class ServiceToolParameter(BaseModel):
    """A parameter for a service tool."""

    model_config = ConfigDict(extra="forbid")

    name: str
    type: str = "string"
    description: str
    required: bool = True
    default: Optional[str] = None


class ConsentRequirement(BaseModel):
    model_config = ConfigDict(extra="forbid")

    operation_kind: str
    token_input_field: str
    target_input_field: str
    min_witnesses: int = 1


class ServiceTool(BaseModel):
    """An MCP-exposed tool for a granted service."""

    model_config = ConfigDict(extra="forbid")

    name: str
    description: str
    parameters: list[ServiceToolParameter] = []
    method: str = "GET"
    path: str
    query_params: Optional[dict[str, str]] = None
    body_template: Optional[dict] = None
    response_path: Optional[str] = None
    requires_consent_token: Optional[ConsentRequirement] = None


class ServiceDefinition(BaseModel):
    """A service that can be granted to agents.

    Loaded from YAML files in ~/.agency/services/.
    """

    model_config = ConfigDict(extra="forbid")

    service: str
    display_name: str
    api_base: str
    description: str = ""
    credential: ServiceCredentialConfig
    usage_example: Optional[str] = None
    tools: list[ServiceTool] = []

    @field_validator("service")
    @classmethod
    def service_name_valid(cls, v: str) -> str:
        if not v.replace("-", "").replace("_", "").isalnum():
            raise ValueError(
                "Service name must be alphanumeric with hyphens or underscores"
            )
        return v


class ServiceGrant(BaseModel):
    """Record that a service has been granted to an agent."""

    model_config = ConfigDict(extra="forbid")

    service: str
    granted_at: str
    granted_by: str

    @field_validator("granted_at")
    @classmethod
    def timestamp_not_empty(cls, v: str) -> str:
        if not v:
            raise ValueError("granted_at must not be empty")
        return v


class AgentServiceGrants(BaseModel):
    """All service grants for an agent.

    Written to ~/.agency/agents/{name}/services.yaml.
    """

    model_config = ConfigDict(extra="forbid")

    agent: str
    grants: list[ServiceGrant] = []
